src/model/CoilUserDefined.cpp

```diff
@@ -0,0 +1,625 @@
+        objectClone.setActuatedComponent(newCoilUserDefined);
+        newCoilUserDefined.setPlantMassFlowRateActuator(objectClone);
+      }
+      //plantOutletTemperatureActuator
+      if (boost::optional<EnergyManagementSystemActuator> object = plantOutletTemperatureActuator()) {
+        auto objectClone = object.get().clone(model).cast<EnergyManagementSystemActuator>();
+        objectClone.setActuatedComponent(newCoilUserDefined);
+        newCoilUserDefined.setPlantOutletTemperatureActuator(objectClone);
+      }
+      //overallModelSimulationProgramCallingManager
+      if (boost::optional<EnergyManagementSystemProgramCallingManager> object = overallModelSimulationProgramCallingManager()) {
+        auto objectClone = object.get().clone(model).cast<EnergyManagementSystemProgramCallingManager>();
+        //clone the program referenced in the callingManager
+        if (boost::optional<EnergyManagementSystemProgram> program = overallSimulationProgram()) {
+          auto cloneProgram = program.get().clone(model).cast<EnergyManagementSystemProgram>();
+          //add cloned program to cloned PCUD
+          newCoilUserDefined.setOverallSimulationProgram(cloneProgram);
+          //add cloned program to cloned programCallingManager
+          objectClone.erasePrograms();
+          objectClone.addProgram(cloneProgram);
+        }
+        newCoilUserDefined.setOverallModelSimulationProgramCallingManager(objectClone);
+      }
+      //modelSetupandSizingProgramCallingManager
+      if (boost::optional<EnergyManagementSystemProgramCallingManager> object = modelSetupandSizingProgramCallingManager()) {
+        auto objectClone = object.get().clone(model).cast<EnergyManagementSystemProgramCallingManager>();
+        //clone the program referenced in the callingManager
+        if (boost::optional<EnergyManagementSystemProgram> program = initializationSimulationProgram()) {
+          auto cloneProgram = program.get().clone(model).cast<EnergyManagementSystemProgram>();
+          //add cloned program to cloned PCUD
+          newCoilUserDefined.setInitializationSimulationProgram(cloneProgram);
+          //add cloned program to cloned programCallingManager
+          objectClone.erasePrograms();
+          objectClone.addProgram(cloneProgram);
+        }
+        newCoilUserDefined.setOverallModelSimulationProgramCallingManager(objectClone);
+      }
+      //ambientZone
+      if (boost::optional<ThermalZone> object = ambientZone()) {
+        //ThermalZone objectClone = object.get().clone(model).cast<ThermalZone>();
+        newCoilUserDefined.setAmbientZone(object.get());
+      }
+      return std::move(newCoilUserDefined);
+    }
+
+    unsigned CoilUserDefined_Impl::airInletPort() const {
+      return OS_Coil_UserDefinedFields::AirConnection1InletNodeName;
+    }
+
+    unsigned CoilUserDefined_Impl::airOutletPort() const {
+      return OS_Coil_UserDefinedFields::AirConnection1OutletNodeName;
+    }
```

Comment:
So Air Connection 2 inlet/outlet nodes are always blank in the idf?

Reply:
yeah, the E+ object allows 2 air connections, but this is a waterToAir, so only 1 air connection.  This is similar to what i did for plantcomponent:Userdefined.  If the object gets used and ppl need the 2nd air connection we could always change it, but thats not gonna be easy.